### ISSUE TITLE
Set cluster-autoscaler.kubernetes.io/safe-to-evict: "true" on sidecar by default (was EmptyDir in istio sidecar prevents cluster autoscaler from evicting a node)

### ISSUE THREAD

**Issue by kyessenov** [istio/istio#19395]
Per https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#what-types-of-pods-can-prevent-ca-from-removing-a-node, presence of local storage emptyDir prevents the cluster autoscaler from evicting nodes with injected pods.

/cc @howardjohn @duderino

**Comment by taisph** [comment]
What's the recommended way to apply the eviction annotation to istio 1.3.8 sidecars?

**Comment by philandstuff** [comment]
A different workaround from using the annotation: call cluster autoscaler with `--skip-nodes-with-local-storage=false`.  This causes the cluster autoscaler to ignore HostPath and EmptyDir for determining which pods are evictable.  (You should check if this is appropriate for your workloads before blindly applying it.)

This is IMO preferable because there may be other reasons that a particular istio-injected pod should be considered unsafe to evict, but injecting the annotation would cause the cluster autoscaler to ignore *all* those criteria.

**Comment by ericbuehl** [comment]
Is there a recommended strategy for making sure the pods in istio-system themselves do not prevent scaling down as they do not have the `safe-to-evict` annotation?

**Comment by seeruk** [comment]
One thing to note here is that on GKE you can't configure CA with `--skip-nodes-with-local-storage=false`. Your only option is to spam the `safe-to-evict` annotation everywhere, which is quite unfortunate.

**Comment by howardjohn** [comment]
Maybe we can just inject the annotation if the user doesn't have emptyDir in their pod spec. Although technically it could cause issues if a later mutating webhook adds an emptyDir

**Comment by seeruk** [comment]
As CA isn't configurable on GKE I opted for developing a mutating admission controller that flips how this behaviour works in a way. If you have a `safe-to-evict` annotation already present (with any value) it leaves it alone, otherwise it adds the `safe-to-evict=true` annotation (excluding kube-system and kube-public namespaces). This way we don't need to spam the annotations anywhere, and the default behaviour is to allow eviction (in our cluster we completely avoid persistent state, so this default makes far more sense).

Unfortunately, it's closed-source right now, but it was pretty simple to develop. After we've battle-tested it a bit more I'll see if I can open source it because it might help sidestep some issues like this one.

**Comment by nrjpoddar** [comment]
Is this still an issue even if you use "memory" emptyDir? We have two emptyDirs usage in proxy, one is memory and the other is not? What is "istio-data" volume used for?

**Comment by thesuperzapper** [comment]
Just wanted to put a solution for those who come across this Issue on Google.

If you are installing the istio with the [istio helm chart](https://github.com/istio/istio/tree/1.4.10/install/kubernetes/helm), you can, use the [`sidecarInjectorWebhook.injectedAnnotations`](https://github.com/istio/istio/blob/1.4.10/install/kubernetes/helm/istio/charts/sidecarInjectorWebhook/values.yaml#L59) value to inject the `safe-to-evict` annotation at the same time as the istio-sidecar.


## Add this to your `values.yaml`:
```yaml
sidecarInjectorWebhook:
  enabled: true
  injectedAnnotations:
    cluster-autoscaler.kubernetes.io/safe-to-evict: "true"
```

**Comment by rmoesbergen** [comment]
> Is this still an issue even if you use "memory" emptyDir? We have two emptyDirs usage in proxy, one is memory and the other is not? What is "istio-data" volume used for?

I checked the source of autoscaler, and the presence of emptyDir (no matter which backend), will count as 'local storage' and prevent eviction.

**Comment by pc-rshetty** [comment]
> Just wanted to put a solution for those who come across this Issue on Google.
> 
> If you are installing the istio with the [istio helm chart](https://github.com/istio/istio/tree/1.4.10/install/kubernetes/helm), you can, use the [`sidecarInjectorWebhook.injectedAnnotations`](https://github.com/istio/istio/blob/1.4.10/install/kubernetes/helm/istio/charts/sidecarInjectorWebhook/values.yaml#L59) value to inject the `safe-to-evict` annotation at the same time as the istio-sidecar.
> 
> ## Add this to your `values.yaml`:
> ```yaml
> sidecarInjectorWebhook:
>   enabled: true
>   injectedAnnotations:
>     cluster-autoscaler.kubernetes.io/safe-to-evict: "true"
> ```

Do you know whats the equivalent on istio operator file when installing with istioctl 
I tried this but validation fails

i tried this config
  meshConfig:
    sidecarInjectorWebhook:
      injectedAnnotations:
        cluster-autoscaler.kubernetes.io/safe-to-evict: "true"
but get error on upgrade (istio 1.6)
2020-12-30T02:56:28.351494Z	info	Error: failed to generate Istio configs from file [/tmp/mgqa-v1-istio-operator-values.yaml], error: validation errors (use --force to override):
failed to unmarshall mesh config: unknown field "sidecarInjectorWebhook" in v1alpha1.MeshConfig

**Comment by pc-rshetty** [comment]
ok works when i specify under values 

```
  values:
    sidecarInjectorWebhook:
      injectedAnnotations:
        cluster-autoscaler.kubernetes.io/safe-to-evict: "true"
```

**Comment by ldemailly** [comment]
Can this be the default?

**Comment by ldemailly** [comment]
John kindly pointed out what I missed; the annotation would evict any pod; while what we want is have the sidecar not prevent eviction but also not  change the logic for the main/app container: if that one does have a volume, that should drive the eviction policy (not the sidecar')

Will file an issue with autoscaler and link here

**Comment by ldemailly** [comment]
closing this one as https://github.com/kubernetes/autoscaler/issues/3947 is really the only safe option outside of duplicating the logic

**Comment by fauzan-n** [comment]
is injectedAnnotations available in istio version 1.2.x ?

**Comment by thesuperzapper** [comment]
For those watching, it looks like [cluster-autoscaler `1.27.0`](https://github.com/kubernetes/autoscaler/releases/tag/cluster-autoscaler-1.27.1) implemented the `cluster-autoscaler.kubernetes.io/safe-to-evict-local-volumes` annotation, you can [read about it in the FAQ](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#what-types-of-pods-can-prevent-ca-from-removing-a-node).

It's a little clunky to use in its current form because we have to list all the local volumes by name, which means that depending on your istio setup (and possibly version) the following annotation may or may not work for you:

```yaml
sidecarInjectorWebhook:
  enabled: true
  injectedAnnotations:
    cluster-autoscaler.kubernetes.io/safe-to-evict-local-volumes: "workload-socket,workload-certs,istio-envoy,istio-data"
```

__Some things to note:__

- you should confirm this list of volumes is comprehensive by looking at what volumes your pods end up with after injection
- setting this annotation with `injectedAnnotations` will make it impossible for specific apps on your cluster to set their own values, because any value they set would be replaced by the sidecar injector
- if your pods have other local-storage volumes, they will not be evictable unless you have also included them in this list 

If `cluster-autoscaler` eventually supports using a wildcard like `"*"`, this might be an alternative that effectively does the 
same thing as setting the overall cluster-autoscaler config `skip-nodes-with-local-storage` to `false`. 

Also, it looks like GKE (who was the main cause of these complaints due to not exposing their cluster-autoscaler configs, has made `skip-nodes-with-local-storage` false by default since [GKE on 1.22](https://cloud.google.com/kubernetes-engine/docs/release-notes#October_27_2021).
